include/execution.hpp

```diff
@@ -360,6 +360,9 @@ namespace _P2300::execution {
             __required_completions) } ->
           __receiver_concepts::__is_valid_completion<remove_cvref_t<_Receiver>>;
       };
+  
+  template<class _Env>
+    using __env_or_void = __if_c<std::is_same_v<_Env, no_env>, void, _Env>;
```

Comment:
I'm curious why this is needed. What happens if you pass `no_env` to the `__awaitable` concept?

Reply:
I guessed that `__awaitable<Sender, Env>` would be specific to that Env, if it is no_env then the promise type is void. If a `no_env` is used for this concept, then it wouldn't satisfy as `await_transform` wouldn't exist.
The following overload will be considered instead of `void*` one
https://github.com/brycelelbach/wg21_p2300_std_execution/blob/2aaf6eb131a1d04c0b3e0dc00f2ce3b960477da1/include/coroutine.hpp#L63-L73

Please correct me if I am wrong.

Reply:
> If a `no_env` is used for this concept, then it wouldn't satisfy as `await_transform` wouldn't exist.

It would, actually! The concept is checking whether a call to `__get_awaiter` with an argument of type `_Promise*` is well-formed. Clearly _this_ overload would not be selected when `_Promise` is `no_env` because `no_env` doesn't have an `await_transform` member.

But! There is another overload of `__get_awaiter` that doesn't require that, and it accepts a `void*` instead of a `_Promise*`. And (here's the key bit) a `_Promise*` pointer is implicitly convertible to a `void*` one.

So the `__awaitable` concept doesn't require you to map `no_env` into a `void` first. You can just pass `no_env` to it.

Reply:
Got it, (and tried it), how silly of me.